File: go-multi-region-s3-crr-kms-cmk/multi-region-s3-crr-kms-cmk/internal/pipeline/multi-region-s3-crr-kms-cmk-pipeline-stack.go
package pipeline

import (
	"github.com/aws/aws-cdk-go/awscdk/v2"
	"github.com/aws/aws-cdk-go/awscdk/v2/awscodecommit"
	"github.com/aws/aws-cdk-go/awscdk/v2/pipelines"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type MultiRegionS3CrrKmsCmkPipelineStackProps struct {
	StackProps awscdk.StackProps
}

func NewMultiRegionS3CrrKmsCmkPipelineStack(scope constructs.Construct, id *string, props *MultiRegionS3CrrKmsCmkPipelineStackProps) {
	stack := awscdk.NewStack(scope, id, &props.StackProps)

	//import codecommit repository
	repo := awscodecommit.Repository_FromRepositoryArn(stack, jsii.String("CodeCommitRepo"), jsii.String("arn:aws:codecommit:us-west-1:442530952648:cdk-pipelines-workshop"))

	pipeline := pipelines.NewCodePipeline(stack, jsii.String("Pipeline"), &pipelines.CodePipelineProps{
		Synth: pipelines.NewShellStep(jsii.String("Synth"), &pipelines.ShellStepProps{
			Input: pipelines.CodePipelineSource_CodeCommit(repo, jsii.String("main"), &pipelines.CodeCommitSourceOptions{}),
			Commands: jsii.Strings(
				"npm install -g aws-cdk && goenv global $GOLANG_16_VERSION",
				"cdk synth",
			),
		}),
	})

	// This is where we add the application stages
	pipeline.AddStage(NewMultiRegionS3CrrKmsCmkStage(stack, jsii.String("Pre-Prod"), awscdk.StageProps{
		Env: &awscdk.Environment{
			Account: props.StackProps.Env.Account,
			Region:  jsii.String("us-west-1"),
		},
	}), &pipelines.AddStageOpts{})

	pipeline.AddStage(NewMultiRegionS3CrrKmsCmkStage(stack, jsii.String("Prod"), awscdk.StageProps{
		Env: &awscdk.Environment{
			Account: props.StackProps.Env.Account,
			Region:  jsii.String("us-west-1"),
		},
	}), &pipelines.AddStageOpts{
		Pre: &[]pipelines.Step{
			pipelines.NewManualApprovalStep(jsii.String("PromoteToProd"), &pipelines.ManualApprovalStepProps{}),
		},
	})
}

File: ts-multi-region-s3-crr-kms-cmk/multi-region-s3-crr-kms-cmk/lib/multi-region-s3-crr-kms-cmk-pipeline-stack.ts
import { Construct } from 'constructs';
import { Stack, StackProps, Stage } from 'aws-cdk-lib';
import * as codecommit from 'aws-cdk-lib/aws-codecommit';
import * as pipelines from 'aws-cdk-lib/pipelines';
import { MultiRegionS3CrrKmsCmkStage } from './multi-region-s3-crr-kms-cmk-stage';

export class MultiRegionS3CrrKmsCmkPipelineStack extends Stack {
    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        const repo = codecommit.Repository.fromRepositoryName(this, 'CodeCommitRepo', 'cdk-pipelines-workshop');

        const pipeline = new pipelines.CodePipeline(this, 'Pipeline', {
            pipelineName: 'MultiRegionS3CrrKmsCmkPipeline',
            
            //how we build and synthesized the application
            synth: new pipelines.CodeBuildStep('SynthStep', {
                input: pipelines.CodePipelineSource.codeCommit(repo, 'main'),
                installCommands: [
                    'npm install -g aws-cdk'
                ],
                commands: [
                    'npm ci',
                    'npm run build',
                    'npx cdk synth'
                ]
            })
        });

        pipeline.addStage(new MultiRegionS3CrrKmsCmkStage(this, 'PreProd', {
            env: {account: props?.env?.account, region: 'us-west-2'}
        }));

        pipeline.addStage(new MultiRegionS3CrrKmsCmkStage(this, 'Prod', {
            env: {account: props?.env?.account, region: 'us-west-2'}
        }), {
            pre: [new pipelines.ManualApprovalStep('PromoteToProd')]
        });
    }
}